### DISCUSSION TITLE
Initialize the material properties at qp

### DISCUSSION THREAD

**Discussion by cooperliu101** [idaholab/moose#18370] [Q&A General] [answered]
Hello, everyone!
In my problem I want to calculate the material property at qp only once at the beginning. initQpStatefulProperties seems to be the solution, but I found initQpStatefulProperties is called only when the material is stateful. 
Is there any function like initQpProperties?

**Comment by hugary1995** [comment, marked as answer]
If you want a material property that persists its value in all time, then it's essentially stateful, right?

I can think of two easy ways of achieving this in moose:

1. You can compute its value in initQpStatefulProperties, and in computeQpProproties, assign the old value into the current property.

2. You can save the value as an AuxVariable and compute it at INITIAL.

I'm sure there are other ways.

**Reply by hugary1995** [reply]
BTW, initQpStatefulProperties is always called, not only on stateful properties.

**Reply by cooperliu101** [reply]
Thanks！
In my moose version （around 2021.4）initQpStatefulProperties is not always called. 
I need to update and check.

according to the latest 
https://github.com/idaholab/moose/blob/4ee8eed2eb28248797bcdc85e998f33f423ff5fa/framework/src/loops/ComputeMaterialsObjectThread.C#L104

https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/framework/src/materials/MaterialPropertyStorage.C#L256

https://github.com/idaholab/moose/blob/66c7cd49507b07bf65d36d128990021fa84148af/framework/src/materials/MaterialBase.C#L108

initQpStatefulProperties seems only called when the material has stateful properties?

**Reply by hugary1995** [reply]
Ah you are absolutely correct. I forget about that.
